pkg/middlewares/compress/compression_handler.go

```diff
@@ -30,10 +35,69 @@ type Config struct {
 	IncludedContentTypes []string
 	// MinSize is the minimum size (in bytes) required to enable compression.
 	MinSize int
+	// Algorithm used for the compression (currently Brotli and Zstandard)
+	Algorithm string
+	// MiddlewareName use for logging purposes
+	MiddlewareName string
+}
+
+const (
+	Brotli    = "br"
+	Zstandard = "zstd"
+)
+
+type compression interface {
+	// Write data to the encoder.
+	// Input data will be buffered and as the buffer fills up
+	// content will be compressed and written to the output.
+	// When done writing, use Close to flush the remaining output
+	// and write CRC if requested.
+	Write(p []byte) (n int, err error)
+	// Flush will send the currently written data to output
+	// and block until everything has been written.
+	// This should only be used on rare occasions where pushing the currently queued data is critical.
+	Flush() error
+	// Close closes the underlying writers if/when appropriate.
+	// Note that the compressed writer should not be closed if we never used it,
+	// as it would otherwise send some extra "end of compression" bytes.
+	// Close also makes sure to flush whatever was left to write from the buffer.
+	Close() error
+}
+type CompressionWriter struct {
+	compression
+	alg string
+}
+
+func NewCompressionWriter(algo string, in io.Writer) (*CompressionWriter, error) {
+	var writer compression = nil
+	switch algo {
+	case Brotli:
+		writer = brotli.NewWriter(in)
+		break
+	case Zstandard:
+		zstdWriter, err := zstd.NewWriter(in)
+		if err != nil {
+			return nil, err
+		}
+		writer = zstdWriter
+		break
+	default:
+		return nil, fmt.Errorf("unknown compression algo: %s", algo)
+	}
+
+	return &CompressionWriter{compression: writer, alg: algo}, nil
+}
```

Comment:
```suggestion
func NewCompressionWriter(algo string, in io.Writer) (*CompressionWriter, error) {
	switch algo {
	case Brotli:
		return &CompressionWriter{compression: brotli.NewWriter(in), alg: algo}, nil

	case Zstandard:
		writer, err := zstd.NewWriter(in)
		if err != nil {
			return nil, err
		}

		return &CompressionWriter{compression: writer, alg: algo}, nil
	default:
		return nil, fmt.Errorf("unknown compression algo: %s", algo)
	}
}
```

Reply:
I disagree with this suggestion. This was the previous version that was suggested, however it's easy to forget to to set the `algo` in the `CompressionWriter` (which was the case in the suggestion).

The version of the code I made account for it, since it won't be the responsibility of the developer adding a new algorithm to the CompressionHandler to not forget to set it.

Also by the time the code reach the creation of the `CompressionWriter` the `algo` has been checked to be valid.

Reply:
I don't share your opinion, Go is not C, we return already.

Reply:
Good point about the `nil` fixing that.

To be honest, I'm trying to have a match pattern like c#: https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/functional/pattern-matching#compare-discrete-values 

One part of the code focused on building the underlying `writer` the other part of the code simply using it to wrap it with only one place setting the `algo`.

Also i'm not so sure what you mean by
> Also break is useless in a switch in Go.
From the go rfc: https://go.dev/ref/spec#Break_statements
> A "break" statement terminates execution of the innermost ["for"](https://go.dev/ref/spec#For_statements), ["switch"](https://go.dev/ref/spec#Switch_statements), or ["select"](https://go.dev/ref/spec#Select_statements) statement within the same function.

Why would it be useless in the case of a `switch` ?

Reply:
> Also i'm not so sure what you mean by

a `case` always breaks, you don't need to add a `break` explicitly.

https://go.dev/ref/spec#Switch_statements

Reply:
> To be honest, I'm trying to have a match pattern like c#

It's not a way to do it in Go, because Go doesn't follow the old rule inherited from C (it was related to memory management), could you apply my suggestion?